src/storage/store/replication/replicator.rs

```diff
@@ -0,0 +1,306 @@
+}
+
+impl Default for ReplicatorSnapshotOptions {
+    fn default() -> Self {
+        ReplicatorSnapshotOptions {
+            interval: 1000,  // Default to taking a snapshot every 1000 blocks
+            max_age: 10_000, // Default to keeping snapshots for 10000 blocks
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Replicator {
+    stores: Arc<ReplicationStores>,
+    snapshot_options: ReplicatorSnapshotOptions,
+}
+
+type ReplicationTransactions = (Vec<proto::Transaction>, Vec<proto::Transaction>);
+
+impl Replicator {
+    pub fn new(stores: Arc<ReplicationStores>) -> Self {
+        Self::new_with_options(stores, ReplicatorSnapshotOptions::default())
+    }
+
+    pub fn new_with_options(
+        stores: Arc<ReplicationStores>,
+        snapshot_options: ReplicatorSnapshotOptions,
+    ) -> Self {
+        Replicator {
+            stores,
+            snapshot_options,
+        }
+    }
+
+    // Fetches a set of system and user transactions that represent the current
+    // state of the given fid range. Start and end are inclusive, i.e. [start, end].
+    pub fn transactions_for_fid_range(
+        &self,
+        height: u64,
+        shard: u32,
+        start_fid: u64,
+        end_fid: u64,
+    ) -> Result<ReplicationTransactions, ReplicationError> {
+        let stores = match self.stores.get(height, shard) {
+            Some(stores) => stores,
+            None => {
+                return Err(ReplicationError::StoreNotFound(
+                    height,
+                    shard,
+                    "No stores found for the given height and shard".to_string(),
+                ));
+            }
+        };
+
+        // TODO: validate fid range
+
+        let mut sys = vec![];
+        let mut user = vec![];
+
+        for fid in start_fid..=end_fid {
+            let sys_tx = build_system_transaction(&stores, fid);
+            let user_tx = build_user_transaction(&stores, fid);
+            sys.push(sys_tx);
+            user.push(user_tx);
+        }
+
+        Ok((sys, user))
+    }
+
+    pub fn handle_post_commit_message(
+        &self,
+        msg: &PostCommitMessage,
+    ) -> Result<(), ReplicationError> {
+        let block_number = match msg.header.height {
+            Some(height) => height.block_number,
+            None => {
+                return Err(ReplicationError::InvalidMessage(
+                    "PostCommitMessage must contain a block number".to_string(),
+                ));
+            }
+        };
+
+        // Clean up old snapshots
+        let expiration_block = block_number.saturating_sub(self.snapshot_options.max_age);
+        self.stores.close_snapshots_below(expiration_block);
+
+        // Check if we can take a snapshot
+        if block_number % self.snapshot_options.interval != 0 {
+            return Ok(());
+        }
+
+        // Open a snapshot
+        self.stores.open_snapshot(block_number, msg.shard_id)
+    }
+}
+
+fn collect_messages_with_page_options<
```

Comment:
Some accounts have a lot of messages (e.g. fids 2 and 3) so we'll need to sync them over multiple requests.

Reply:
Good call. We've chatted about this, and we're deferring on implementing handling for this PR.